sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubProducerClient.cs

```diff
@@ -440,13 +458,28 @@ public virtual Task<string[]> GetPartitionIdsAsync(CancellationToken cancellatio
                                             CancellationToken cancellationToken = default)
         {
             Argument.AssertNotNull(eventBatch, nameof(eventBatch));
-            AssertSinglePartitionReference(PartitionId, eventBatch.SendOptions.PartitionKey);
+            AssertSinglePartitionReference(eventBatch.SendOptions.PartitionId, eventBatch.SendOptions.PartitionKey);
+
+            // Determine the transport producer to delegate the send operation to.  Because sending to a specific
+            // partition requires a dedicated client, use (or create) that client if a partition was specified.  Otherwise
+            // the default gateway producer can be used to request automatic routing from the Event Hubs service gateway.
+
+            TransportProducer activeProducer;
+
+            if (String.IsNullOrEmpty(eventBatch.SendOptions.PartitionId))
+            {
+                activeProducer = GatewayProducer;
+            }
+            else
+            {
+                activeProducer = PartitionProducers.GetOrAdd(eventBatch.SendOptions.PartitionId, id => Connection.CreateTransportProducer(id, Options));
```

Comment:
Nit: Can we move this logic to a separate `getActiveProducer` (or something similar) method which can be re-used from `SendAsync(IEnumerable{EventData}, SendOptions, CancellationToken)` ?

Reply:
I don't think there's much value in that; we're talking about moving a single line of uncomplicated code to a separate area.  I dislike that the detail would be hidden and require a context-shift for those reading the code.  I also don't see a need to take the potential hit of a method call or marking it up for aggressive inlining.